complex/cargo/list/src/lib.rs

```diff
@@ -37,228 +49,778 @@ pub struct ListManager {
-        let conn = self.store.read_connection();
-        let mut stmt = conn.prepare(sql).unwrap();
-        let mut item_iter = stmt.query_map(&[&label.name], |row| {
-            let uuid: String = row.get(0);
-            Item {
-                uuid: uuid.clone(),
-                name: row.get(1),
-                due_date: row.get(2),
-                completion_date: row.get(3),
-                labels: self.fetch_labels_for_item(&uuid)
-            }
-        }).unwrap();
+    }
 
-        let mut item_list: Vec<Item> = Vec::new();
-        while let Some(result) = item_iter.next() {
-            if let Some(i) = result.ok() {
-                item_list.push(i);
+    pub fn fetch_item(&self, uuid: &Uuid) -> Result<Option<Item> , list_errors::Error>{
+        let query = r#"[:find ?eid, ?uuid, ?name
+            :in ?uuid
+            :where
+            [?eid :item/uuid ?uuid]
+            [?eid :item/name ?name]
+        ]"#;
+        let result = Arc::clone(&self.store).query_args(query, &[&(&"?uuid".to_string(), &uuid)])?;
+        if let QueryResults::Rel(rows) = result {
+            if !rows.is_empty() {
+                let row = &rows[0];
+                Ok(Some(Item{
+                    id: row[0].clone().to_inner(),
+                    uuid: row[1].clone().to_inner(),
+                    name: row[2].clone().to_inner(),
+                    due_date: self.fetch_date_for_item("due_date", &uuid).unwrap_or(None),
+                    completion_date: self.fetch_date_for_item("completion_date", &uuid).unwrap_or(None),
+                    labels: self.fetch_labels_for_item(&uuid).unwrap_or(vec![]),
+                }))
+            } else {
+                Ok(None)
             }
+        } else {
+            Ok(None)
         }
-        item_list
     }
 
-    pub fn fetch_item(&self, uuid: &String) -> Option<Item> {
-        let sql = r#"SELECT uuid, name, due_date, completion_date FROM items WHERE uuid=?"#;
-
-        let conn = self.store.read_connection();
-        let mut stmt = conn.prepare(sql).unwrap();
-        let mut item_iter = stmt.query_map(&[uuid], |row| {
-            Item {
-                uuid: row.get(0),
-                name: row.get(1),
-                due_date: row.get(2),
-                completion_date: row.get(3),
-                labels: self.fetch_labels_for_item(uuid)
+    fn fetch_date_for_item(&self, attr: &str, item_id: &Uuid) -> Result<Option<Timespec>, list_errors::Error> {
+        let query = format!(r#"[:find ?{0}
+            :in ?uuid
+            :where
+            [?eid :item/{0} ?{0}]
+            [?eid :item/uuid ?uuid]
+        ]"#, attr);
+        let result = Arc::clone(&self.store).query_args(&query, &[&(&"?uuid".to_string(), &item_id)])?;
+        if let QueryResults::Rel(rows) = result {
+            if !rows.is_empty() {
+                let row = &rows[0];
+                let date: Option<Timespec> = row[0].clone().to_inner();
+                Ok(date)
```

Comment:
Why is this whole block not:

```
Ok(rows.first().map(|ts| ts.to_inner())))
```
?

But you should be using a scalar query here, binding the attribute rather than doing text substitution (!!!!), and putting the most specific clause first:

```
[:find ?val .
 :in ?a ?uuid
 :where
 [?eid :item/uuid ?uuid]
 [?eid ?a ?val]]
```

Reply:
So, I struggled with binding an attribute, passing in a `NamespaceKeyword` as an arg resulting in an error: `CC known empty: Type mismatch: var(?a) can't be ValueTypeSet({Ref}), because it's already ValueTypeSet({Keyword})`.

If it's expecting a ref, am I supposed to do a query to get the ref of the attribute first and then another to use that attribute as an arg? That makes no sense, so what am I doing wrong?